### PR TITLE
Remove legacy version, improve filter

### PR DESCRIPTION
**Pull request by thatch**
I'm reasonably sure this gives the same result, only with much less work.  This work happens in the main thread (blue/green 2nd row) and is noticeable with hot cache or extremely low latency:

Before
old
After
new

### CODE REVIEW

**Comment by thatch** [comment]
Needs a little more work on the error message (which doesn't have any tests), and some sort of validation that this still matches what "popular" tools do on at least non-backtracking solves.

File: hdeps/compatibility.py [hunk 1]
```diff
@@ -19,55 +18,75 @@ def find_best_compatible_version(
...
+        return verdict
+
+    possible: List[Version] = []
+
+    specifier_matched = False
+    with kev("reverse"):
+        rev = reversed(project.versions.keys())
+    with kev("initial filter"):
+        for version in req.specifier.filter(rev):
+            specifier_matched = True
```
Comment (zsimic) [inline hdeps/compatibility.py:60]: `specified_matched` seems unused

File: hdeps/compatibility.py [hunk 2]
```diff
@@ -19,55 +18,75 @@ def find_best_compatible_version(
...
+            if requires_python_match(version):
+                # Only keep the first one
+                possible.append(version)
+                break
```
Comment (zsimic) [inline hdeps/compatibility.py:64]: Would this be the same as this?
```
possible = any(v for v in req.specifier.filter(rev) if requires_python_match(v))
```

Comment (zsimic) [reply hdeps/compatibility.py:64]: Oops, not quite, more like a `possible.extend()` but then it's less readable

Comment (thatch) [reply hdeps/compatibility.py:64]: Yeah, and I'm pretty sure the magic around prerelease versions got lost.  It's about this point that it became a draft PR, clearly needs more testing.

**Review by zsimic — APPROVED** [review]

**Pull request merged**